### DISCUSSION TITLE
How can I trigger a function once the browser tab is closed?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#57987] [Help] [answered]
### Summary

This might be the wrong of thinking... but I am trying to use the localStorage to keep a sessionId....  then check if the sessionId still valid on the browser, and if not, then trigger a function....

The localStore is supposed to eliminate all information once the tab is closed, but if the the tab is closed, how can you check for an item at the localStore?

What would be the best way to do this?

So far I got this, but this isn't doing anything:

```
useEffect(() => {
    localStorage.setItem("sessionId", conversationId);
  }, []);

useEffect(() => {
    if (tab) {
      const sessionId = localStorage.getItem("sessionId");
      const intervals = setInterval(() => {
        if (sessionId === null) {
          if (chatToServer.length > 0) {
            emailChatHistory();
            setTab(false);
          }
        }
        if (tab === false) {
          clearInterval(intervals);
          setTab(true);
        }
      }, 1000);
    }
  });

```

^^^ Oh yes, besides this doing nothing when the tab is closed, it also keeps changing the value of the sessionId on every refresh, which doesn't help keep track of the session..

**Comment by Elindo586** [comment]
I tried this too.... not working...

```
if (tab) {
    if (typeof window !== "undefined") {
      const sessionId = localStorage.getItem("sessionId");
      const intervals = setInterval(() => {
        if (sessionId === null) {
          if (chatToServer.length > 0) {
            window.onbeforeunload = function () {
              emailChatHistory();
              setTab(false);
            };
          }
          if (tab === false) {
            clearInterval(intervals);
            setTab(true);
          }
        }
      }, 1000);
    }
  }
```

**Comment by icyJoseph** [comment]
Local Storage is preserved. I think you are talking about `Session Storage`?

**Reply by icyJoseph** [reply]
@Elindo586 what is it you are trying to do? Where do you get the session-id, and how do you intend on refreshing it? Why would the tab being closed matter? 

I think the best way is the `visibilitychange` event, but the code you've shared is confusing.

- https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event

**Reply by Elindo586** [reply]
Yes, the `visibilitychange `might work better... gonna try to implement it later today..  

And yes, it is also the `sessionStorage` that I was thinking of...

The `conversationId` is set as the window is open, and if there is a conversation going then it takes the id of the conversation coming from the Redux store...   so far it works to keep track of the conversation, but yes, using the `sessionStorage` the Id keeps changing on refresh, so if I am gonna use that number I gotta make sure it doesn't change on a refresh as well...  OOOHHH.. and just now testing it, and refreshing the page reloads a new conversation on everything.... 

The code is long and didn't want to post it all.... but here is the link if you want to look at it.
https://github.com/Elindo586/nextprojectd/blob/main/components/chat/message-input.js

**Reply by icyJoseph** [reply]
Wouldn't it make more sense to use `localStorage` then? Since that preserves the data across reloads?

**Reply by Elindo586** [reply]
My goodness.... gonna have to re-work a lot in here...

For what I understand `sessionStore `  is able to maintain information during page refresh...  apparently I am doing something in my code that keeps putting a new id on each reload.

Redux and the entire code doesn't seem to survive a page reloads as the conversation goes away from my input display, and the system starts a new conversation chain.

^^^ this is going to be a battle for a different day more likely.

For today I am looking to do the function call on tab closed....  gonna try some different code as you suggested above.

**Reply by Elindo586** [reply]
Hey... you asked:
Why would the tab being closed matter?

^^^  that is because I want to send chat history out if there was a conversation before closing the tab.

**Reply by icyJoseph** [reply]
Aha,

Ok, so you only care about the one tab, not other tabs, nor people closing the tab entirely and then later on coming back.

So session storage does seem like a good candidate. 

So let's stick to session storage. Now what you want is to do some work, before the tab is destroyed. That's kind of tricky. 

First, everything that's within the JS context of the page, will be destroyed when you reload. For example the redux store. There's ways to persist a redux store, but let's not get into that.

- You should check if there's an already stored id in the Session Storage, before setting a new one.
- When the visibility goes to hidden, you could try to send the information you need

**Reply by Elindo586** [reply]
This code below works if... switches tab, minimize the screen....  but doesn't work on reloads or close tab.. (the documents said it would work on tab close)..

Another thing is that when switching tabs, or minimizing the screen works on triggering a function... when I type another line the code breaks at the openai api for some reason... it seems like its killing all data and I need to put it back in there fast before the user takes any quick actions... according to the documents the `"hidden"`  `visibilityState `  should be treated as end of session,  but that isn't exactly what is happening on an ongoing api connection with data..


```
useEffect(() => {
    document.addEventListener("visibilitychange", () => {
      if (document.visibilityState === "hidden") {
        if (chatToServer.length > 0) {
          emailChatHistory();
        }
      }
    });
  });

```


Now....  I would be happy on just sending chuncks of information on page reloads and tab closes... doesn't matter if it starts a new chain as long as I could have the chuncks of the previous conversation.

**Reply by icyJoseph** [reply]
Could you share the `emailChatHistory` function? I guess it does a `fetch(/* */)` call? You can pass a `keepAlive` option to fetch, and that should do the trick I think~~

 - https://developer.mozilla.org/en-US/docs/Web/API/fetch#keepalive

**Reply by Elindo586** [reply]
I am gonna try the `keepAlive` option..  

Now I do have a way to send chat history when the user press the close button... if they press it, and they might do it to keep navigating the site without the chat window open... 

The other feature I was looking for was to send chat history at tab close (which is not able to do right now) because that is another very likely situation to send information..... at tab minimize or change might not be the best options to send information, but for lack of alternatives would be something to take :p  

The `emailChatHistory` function is below... it just sends active data via email

```
const emailChatHistory = async () => {
    await fetch("/api/chat-email", {
      method: "POST",
      headers: {
        Accept: "application/json, text/plain, */*",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(chatToEmailServer),
    }).then((res) => {
      console.log("response received");
      if (res.status === 200) {
        console.log("response succeeded");
      }
    });
  };

```

**Comment by Yefee8** [comment]


**Reply by Elindo586** [reply]
So I just saw this post.... question... Does it work on Next.js???   I ask because Next.js doesn't have  a window to track, and BeforeUnload tracks changes on the window.

**Reply by icyJoseph** [reply]
I think you are misunderstanding a few things here. The window exists in code that runs client side.

- https://github.com/streamich/react-use/blob/master/src/useBeforeUnload.ts

If we look at that function implementation, a handler is defined, and within a use-effect, which runs on the browser, the handler is attached to the window, `beforeunload` event. 

However, that's not necessarily good. Some tools will complain about that when checking your site.

It is better to use the visibility change event:

- https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event#examples

**Reply by Elindo586** [reply]
Oh it's been a while now.  I think I did try the visibility change, and I think I did work when hiding the tab, but not when closing the tab.

It isn't too much of a big deal, but the function I want to run, I want it at the closing tab of the browser, that way I can get all the information of a chat to be sent via email.

Right now it is working on closing the chat `div`,  but I would think many users don't close the chat and rather just close the browser.

**Reply by Vendjin** [reply]
Tell me, does this work in such a case?
If the tab is closed, send a request to the API, and if you reload the page, then no action is required to send to the API?

**Reply by diasmashikov** [reply]
Any updates?

**Reply by Vendjin** [reply]
no not any, but only closure

**Comment by Elindo586** [comment, marked as answer]
I got this one working with the `beforeunload `method 

Went some like this:

  ```
useEffect(() => {
    window.addEventListener("beforeunload", () => {
      if (chatToEmailServer.chatToServer.length > 0) {
        emailChatHistory();
        setFirstMsg(true);
      }
    });
  }, [chatToEmailServer.chatToServer, firstMsg]);
```

^^^  This is working, but it gives me warning on the code for dependencies.  

Trying to fix the dependencies error is something in the backburner because include the `emailChatHistory()` in the dependencies, then I have to dig all through de code and fix communication with the API and a few other things....  basically it opens another chain of errors that I couldn't figured out.... so, until it breaks, or when I got time, I might change the code later.
